feat(NOTIFY-866): add account syncing

mathieuartu opened the pull request:
## **Description**

This PR adds [account syncing](https://www.figma.com/board/sx26FtZcQc0zs0z5hpdqBh/Account-Syncing-Solution?t=5Oy00gbHLnoPqyYP-0) to MetaMask.
As part of profile syncing, with this new feature, users will be able to synchronize all of their accounts and accounts' names across all of their devices.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/27060?quickstart=1)

## **Related issues**

Fixes:

[NOTIFY-866](https://consensyssoftware.atlassian.net/jira/software/projects/NOTIFY/boards/616?assignee=712020%3A5843b7e2-a7fe-4c45-9fbd-e1f2b2eb58c2&selectedIssue=NOTIFY-866)

## **Manual testing steps**

1. Log in with SRP
2. Activate profile syncing in preferences
3. Add new accounts and rename others
4. Using another browser, log in with SRP
5. Activate profile syncing in preferences
6. Watch the magic happen!

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


[NOTIFY-866]: https://consensyssoftware.atlassian.net/browse/NOTIFY-866?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mathieuartu committed:
feat(NOTIFY-866): add account syncing

mathieuartu committed:
fix: yarn.lock & keyring-controller patch

mathieuartu committed:
fix: lavamoat policies

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: add back isUnlock

mathieuartu committed:
feat: add unit tests

metamaskbot committed:
Update LavaMoat policies

mathieuartu committed:
fix: mock hook

mathieuartu committed:
fix: add temporary preview package

mathieuartu committed:
fix: yarn.lock

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: update preview package

mathieuartu committed:
fix: await in e2e tests

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: update useAccountSyncingEffect deps

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
test: add coverage

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: update JSDoc and mimic existing hook logic

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

Prithpal-Sooriya commented on .yarn/patches/@metamask-keyring-controller-npm-17.2.0-667fdd6864.patch:
```diff
@@ -1,5 +1,5 @@
-index 5a6217eaed16fdfe7f1ad693871f85320bd6b421..69bdf1a9155497e37fc58db7bbc74597fd543535 100644
+index 1caaa7e433c4597d2554227536e87f16ca9ea117..29c11adb88149d3d6aef5abc049c19dedab71a91 100644
 --- a/package.json
 +++ b/package.json
```
Whats up with this patch change?

mathieuartu replied:
I kept the original patch, but updated it to match the bumped version of keyring-controller

Prithpal-Sooriya commented on app/scripts/metamask-controller.js:
```diff
@@ -1472,10 +1472,14 @@ export default class MetamaskController extends EventEmitter {
         this.metaMetricsController.state.participateInMetaMetrics ?? false,
       state: initState.UserStorageController,
+      env: {
+        isAccountSyncingEnabled: true,
```
Can't wait!!

(no real opinion on this). Was thinking we have this turned off on this PR, and then we can make a subsequent PR to turn on this feature.

But we can keep as is otherwise (just notify extension team so QA doesn't get nicely surprised)

Prithpal-Sooriya commented on app/scripts/metamask-controller.js:
```diff
@@ -1472,10 +1472,14 @@ export default class MetamaskController extends EventEmitter {
         name: 'UserStorageController',
         allowedActions: [
           'KeyringController:getState',
+          'KeyringController:addNewAccount',
```
Nice addition to the allowed actions and events.

I'm thinking if it is worth adding comments to indictate which actions/events are for what features.

Mostly because this controller will start to have a large set of actions & events it will listen to with all the syncing features.

Meh, we can do this later.

Prithpal-Sooriya commented on app/scripts/metamask-controller.js:
```diff
@@ -3877,6 +3888,10 @@ export default class MetamaskController extends EventEmitter {
         userStorageController.setIsProfileSyncingEnabled.bind(
           userStorageController,
         ),
+      syncInternalAccountsWithUserStorage:
+        userStorageController.syncInternalAccountsWithUserStorage.bind(
+          userStorageController,
+        ),
```
Nice binding! Do we will be using this in the UI? 😄 
I will probably do the same for network syncing.

Discuss - should we have 1 "sync" function fires off all sync features, or have this for granularity?

Lets stick with this for now.

Prithpal-Sooriya commented on package.json:
```diff
@@ -344,7 +343,7 @@
     "@metamask/post-message-stream": "^8.0.0",
     "@metamask/ppom-validator": "0.34.0",
-    "@metamask/profile-sync-controller": "^0.2.1",
+    "@metamask/profile-sync-controller": "npm:@metamask-previews/profile-sync-controller@0.5.0-preview-c1e764d",
```
I never know module aliasing was a thing!
Before I used to always find/replace all imports, this is much nicer ❤️

Prithpal-Sooriya commented on test/e2e/tests/notifications/mocks.ts:
```diff
@@ -20,18 +20,18 @@ type MockResponse = {
 
   // Storage
-  mockAPICall(server, StorageMocks.getMockUserStorageGetResponse());
+  mockAPICall(server, await StorageMocks.getMockUserStorageGetResponse());
```
Nice.

@matteoscurati has a PR (unsure if merged) that resolves all this async mocks.

Context - some mocks are now async due to encryption now being async (and how our mocks use realistic encrypted values).

Prithpal-Sooriya replied:
These changes will be available in this PR:
https://github.com/MetaMask/metamask-extension/pull/26987

Prithpal-Sooriya commented on ui/contexts/metamask-notifications/metamask-notifications.tsx:
```diff
@@ -45,6 +46,8 @@ export const MetamaskNotificationsProvider: React.FC = ({ children }) => {
     [isProfileSyncingEnabled, isNotificationsEnabled],
   );
 
+  useAccountSyncingEffect();
```
CLEAN!

(Future) - I don't like this context, I'd rather we replace all this with hooks too.

Prithpal-Sooriya commented on ui/hooks/metamask-notifications/useProfileSyncing.test.tsx:
```diff
@@ -79,4 +82,88 @@ describe('useProfileSyncing', () => {
   });
+
+  it('should dispatch account syncing when conditions are met', async () => {
+    store = mockStore({
+      metamask: {
+        isProfileSyncingEnabled: true,
+        isSignedIn: true,
+        isUnlocked: true,
+        useExternalServices: true,
+        participateInMetaMetrics: false,
+        internalAccounts: {
+          accounts: {
+            '0x123': {
+              address: '0x123',
+              id: 'account1',
+              metadata: {},
+              options: {},
+              methods: [],
+              type: 'eip155:eoa',
+            },
+          },
+        },
+      },
+    });
+
+    store.dispatch = jest.fn().mockImplementation((action) => {
+      if (typeof action === 'function') {
+        return action(store.dispatch, store.getState);
+      }
+      return Promise.resolve();
+    });
+
+    jest.clearAllMocks();
```
Since these tests use similar mocks, can we refactor and create an `arrangeMocks` function?

(long term) I kind of want to clean up this file too.
- Rather than 1 big describe, lets create multiple describe blocks for the separate tests
- Refactor & Reused the store and dispatch mocks.
- (minor) Not a fan of beforeEach, it's always felt flakey. I prefer arrangeMock fns. But IIWIW.

Prithpal-Sooriya commented on ui/hooks/metamask-notifications/useProfileSyncing.ts:
```diff
@@ -112,3 +118,63 @@ export function useSetIsProfileSyncingEnabled(): {
+export const useAccountSyncing = () => {
+  const dispatch = useDispatch();
+
+  const [isLoading, setLoading] = useState<boolean>(false);
+  const [error, setError] = useState<unknown>(null);
```
Since we never use the isLoading/isError, should we keep it?

Prithpal-Sooriya commented on ui/hooks/metamask-notifications/useProfileSyncing.ts:
```diff
@@ -112,3 +118,63 @@ export function useSetIsProfileSyncingEnabled(): {
+    } catch (e) {
+      log.error(e);
+      setError(e instanceof Error ? e.message : 'An unexpected error occurred');
+      throw e;
```
Q - We do this with our other hooks. I'm not entirely convinced, should hooks throw?

Prithpal-Sooriya commented on ui/store/actions.ts:
```diff
@@ -5370,6 +5370,34 @@ export function fetchAndUpdateMetamaskNotifications(): ThunkAction<
+ * Synchronizes accounts data with user storage between devices.
+ *
+ * This function sends a request to the background script to sync accounts data and update the state accordingly.
+ * Upon success, it dispatches an action with type `SYNC_INTERNAL_ACCOUNTS_WITH_USER_STORAGE` to update the Redux state.
```
Hmm no I don't think this is true.

@matteoscurati confirm? I think these are old comments that were never removed.

We can create a cleanup ticket.

mathieuartu committed:
fix: remove loading state and add arrangeMocks function

mathieuartu committed:
fix: remove throw and update jsdoc

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: conflict

mathieuartu committed:
fix: conflicts

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
chore: bump packages to latest and remove preview dependencies

mathieuartu committed:
fix: yarn.lock

mathieuartu committed:
chore: bump @metamask/snaps-rpc-methods version

gambinish committed:
chore: Update `asset-list` to Typescript, remove PropTypes (#26952)

## **Description**

I'm working on Token sorting, and since I was in the area, I took the
opportunity to update the components I was touching to Typescript. This
PR handles those changes, and I'll work on the actual sorting logic in a
separate branch that I will merge with this one.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/26952?quickstart=1)

## **Related issues**

First step of:
https://consensyssoftware.atlassian.net/browse/MMASSETS-356

## **Manual testing steps**

Nothing should be different visually or functionally. ~~I left a comment
on a suspicious snapshot test where I can't explain the diff, other than
the original snapshot not being accurate based on the mocks~~ 👈 no
longer relevant.

## **Screenshots/Recordings**

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Brian Bergeron <brian.e.bergeron@gmail.com>

jclancy93 committed:
fix: change asset picker button to buttonbase to fix text theming (#27127)

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->
The asset picker in dark mode previously showed the down icon and token
symbol in their light mode variants, instead of dark mode. This is
because `Button` defaults to the primary variant which hard codes the
theme as light mode for all descendant elements. This PR changes the
AssetPicker main component from Button to ButtonBase since it does not
default to light mode

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/27127?quickstart=1)

## **Related issues**

Fixes: #27046 

## **Manual testing steps**

1. Change to dark mode
2. Open Send Page
3. Asset symbol and down carat should both render as white
4. Switch to light mode
5. Asset symbol and down carat should both render as black

## **Screenshots/Recordings**

### **Before**

<img width="1624" alt="image"
src="https://github.com/user-attachments/assets/69e086b1-daec-46d9-b299-74434d58a616">


### **After**

<img width="1624" alt="image"
src="https://github.com/user-attachments/assets/709898ae-9766-4902-a089-9183bac8fcc1">


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

mathieuartu commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: update package.json & yarn.lock

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: revert package version

mathieuartu committed:
fix: update package.json & yarn.lock

mathieuartu committed:
fix: snaps-sdk related tests

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
feat: add analytics

mathieuartu committed:
fix: yarn.lock

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing

mathieuartu committed:
fix: revert snaps changes

mathieuartu committed:
fix: yarn.lock

Prithpal-Sooriya commented on .yarn/patches/@metamask-keyring-controller-npm-17.1.1-098cb41930.patch:
Nice that we don't need this patch anymore. Can you confirm that this is correct?

ccharly replied:
I double-checked, and nope we do not need this patch anymore>

Since this PR uses the version [17.2.1](https://github.com/MetaMask/core/blob/main/packages/keyring-controller/CHANGELOG.md#1721) of the `keyring-controller`, the package is now built using `ts-bridge` which solves those issues!

matteoscurati reviewed (approved):
LGTM

mathieuartu committed:
Merge branch 'develop' into feat/add_account_syncing
